From 644c74862bb0015564c38288203cc50371729427 Mon Sep 17 00:00:00 2001
From: Michael Matz <matz@suse.de>
Date: Wed, 12 Jun 2013 17:03:42 +0200
Subject: [PATCH 157/169] aarch64: Fix 32bit TST

A 32bit TST insn (aka ANDS) was using the 64bit form to
set the flags, hence it was regarding bit 63 as sign bit,
not bit 31.
---
 target-arm/translate-a64.c | 9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

Index: qemu-1.6.0+dfsg/target-arm/translate-a64.c
===================================================================
--- qemu-1.6.0+dfsg.orig/target-arm/translate-a64.c	2013-11-05 22:23:48.031351619 +0000
+++ qemu-1.6.0+dfsg/target-arm/translate-a64.c	2013-11-05 22:23:48.023351639 +0000
@@ -755,15 +755,18 @@
         break;
     }
 
+    tcg_temp_free_i64(tcg_op2);
+
     if (is_32bit) {
         tcg_gen_ext32u_i64(tcg_dest, tcg_dest);
     }
 
     if (setflags) {
-        gen_helper_pstate_add(pstate, pstate, tcg_dest, cpu_reg(31), tcg_dest);
+	if (is_32bit)
+	  gen_helper_pstate_add32(pstate, pstate, tcg_dest, cpu_reg(31), tcg_dest);
+	else
+	  gen_helper_pstate_add(pstate, pstate, tcg_dest, cpu_reg(31), tcg_dest);
     }
-
-    tcg_temp_free_i64(tcg_op2);
 }
 
 static void setflags_add(bool sub_op, bool is_32bit, TCGv_i64 src,
